File: LeetCode/0053_maximum-subarray/20220528.ex
defmodule Solution do
  @spec max_sub_array(nums :: [integer]) :: integer
  def max_sub_array(nums) do
    nums
      |> Enum.reverse()
      |> Enum.reduce([], fn num, dp ->
          case dp do
            [] ->
              [num]
            _ ->
              [next_optimal | remainings] = dp
              [max(num, num + next_optimal), next_optimal] ++ remainings
          end
        end)
      |> Enum.reduce(&(max(&1, &2)))
  end
end

File: LeetCode/Medium/0053_maximum-subarray/20220528.py
class Solution:
    def maxSubArray(self, nums: list[int]) -> int:
        dp = nums[:]
        
        for i in range(len(dp) - 2, -1, -1):
            dp[i] = max(dp[i], dp[i] + dp[i + 1])
            
        return max(dp)
